pydm/widgets/base.py

```diff
@@ -498,6 +524,115 @@ def force_redraw(self):
         """
         self.update()
 
+    @pyqtSlot(dict)
+    def setRule(self, payload):
+        """
+        Callback called when a rule has a new value for a property.
+
+        Parameters
+        ----------
+        payload : dict
+            Dictionary containing the rule name, the property to be set and the
+            new value.
+
+        Returns
+        -------
+        None
+        """
+        name = payload.get('name', '')
+        prop = payload.get('property', '')
+        value = payload.get('value', None)
+
+        if prop not in PyDMWidget.RULE_PROPERTIES:
+            logger.error('Error at Rule: %s. %s is not part of this widget properties.',
+                         name, prop)
+            return
+
+        method_name, data_type = PyDMWidget.RULE_PROPERTIES[prop]
+        method = getattr(self, method_name)
+        method(value)
+
+    @pyqtProperty(str)
+    def rules(self):
+        """
+        JSON-formatted list of dictionaries, with rules for the widget.
+
+        Returns
+        -------
+        str
+        """
+        return self._rules
+
+    @rules.setter
+    def rules(self, new_rules):
+        """
+        JSON-formatted list of dictionaries, with rules for the widget.
+
+        Parameters
+        ----------
+        new_rules : str
+
+        Returns
+        -------
+        None
+        """
+        if new_rules != self._rules:
+            self._rules = new_rules
+
+            # Let's clean up the current actions
+            # and terminate all the action threads
+            # that are there...
+            if len(self._rules_objs) != 0:
+                for ao in self._rules_objs:
+                    ao.requestInterruption()
+                    ao.quit()
+                    ao.wait()
+                self._rules_objs = []
+            try:
+                rules_list = json.loads(self._rules)
+            except JSONDecodeError:
+                logger.error('Invalid format for Rules')
+                return
+
+            for ro in rules_list:
+                self._rules_objs.append(RulesEngine(ro))
```

Comment:
Concerned that this will be a thread per rule....

Reply:
It is! So far we had no problem with having multiple QThreads running... the Modbus plugin for instance runs fine with 200+ channels connected and no issues so far... If this becomes a problem in the future we can think about creating a `Rules` thread that handles all the rules sequentially but I think that this would be bad since they would not act in parallel.